### ISSUE TITLE
y_lock triggered errors

### ISSUE THREAD

**Issue by Misiur** [pawn-lang/YSI-Includes#137]
Hello, I don't remember if y_lock was ported from YSI 3 but there's an user who wants to use it and maybe a few little tweaks will fix this.

Code (tested on Zeex's compiler)

```pawn
#define BIND_IP MAKE_LOCK_IP(127.0.0.1)
#include <YSI\y_lock>

#include <a_samp>

main() {}

public OnGameModeInit()
{
    return 1;
}
```

Errors:
1. https://github.com/Misiur/YSI-Includes/blob/f5b9a39cfb9c1b56a77eba7abc00d951f6b86607/YSI_Server/y_lock.inc#L72 - it seems preprocessor doesn't shortcircuit && (nor ||) statements. Results in `_DEBUG` is not defined. Simple enough, just expand it and repeat the else clause
2. https://github.com/Misiur/YSI-Includes/blob/f5b9a39cfb9c1b56a77eba7abc00d951f6b86607/YSI_Internal/y_shortfunc.inc#L275 - hook is not properly, well, hooked? After running through preprocessor, the result in .lst is:
```raw
#file "D:\Dev\PAWN\ugrp\pawno\include\YSI\..\YSI_Server\..\YSI_Coding\..\YSI_Core\..\YSI_Internal\..\YSI_Coding\y_hooks.inc"
#line 73

#line 75
	#endinput
#file "D:\Dev\PAWN\ugrp\pawno\include\YSI\..\YSI_Server\..\YSI_Coding\..\YSI_Core\..\YSI_Internal\y_shortfunc.inc"
#line 275
hook AMX_OnScriptInit()
{
```

Not important errors, but for sake of consistency:
```
D:\Dev\PAWN\ugrp\pawno\include\YSI\..\YSI_Server\..\YSI_Coding\..\YSI_Core\..\YSI_Internal\y_shortfunc.inc(275) : error 010: invalid function or declaration
D:\Dev\PAWN\ugrp\pawno\include\YSI\..\YSI_Server\..\YSI_Coding\..\YSI_Core\..\YSI_Internal\y_shortfunc.inc(278) : error 001: expected token: ";", but found "-identifier-"
D:\Dev\PAWN\ugrp\pawno\include\YSI\..\YSI_Server\..\YSI_Coding\..\YSI_Core\..\YSI_Internal\y_shortfunc.inc(281) : error 001: expected token: ";", but found "-identifier-"
D:\Dev\PAWN\ugrp\pawno\include\YSI\..\YSI_Server\..\YSI_Coding\..\YSI_Core\..\YSI_Internal\y_shortfunc.inc(282) : error 010: invalid function or declaration
D:\Dev\PAWN\ugrp\pawno\include\YSI\..\YSI_Server\..\YSI_Coding\..\YSI_Core\..\YSI_Internal\y_shortfunc.inc(285) : error 054: unmatched closing brace ("}")
```

This one is a little weird. I tried to check if "normal" callbacks will work, so I did

```pawn
diff --git a/YSI_Internal/y_shortfunc.inc b/YSI_Internal/y_shortfunc.inc
index 89cfbaf..4efbe63 100644
--- a/YSI_Internal/y_shortfunc.inc
+++ b/YSI_Internal/y_shortfunc.inc
@@ -272,6 +272,14 @@ stock F@()
 
 #if defined YSI_LOCK_MODE
 	#include "..\YSI_Coding\y_hooks"
+
+	hook OnGameModeInit()
+	{
+		print("FUCK");
+
+		return 1;
+	}
+
 	hook OnScriptInit()
 	{
 		new
```

and the resulting .lst:

```raw
#line 75
	#endinput
#file "D:\Dev\PAWN\ugrp\pawno\include\YSI\..\YSI_Server\..\YSI_Coding\..\YSI_Core\..\YSI_Internal\y_shortfunc.inc"
#line 275

hook ScriptInit_OnGameModeInit() <_ALS : _ALS_go>
{
	print("FUCK");

	return 1;
}

hook AMX_OnScriptInit()
{
```
It seems none of the hooks are working there. 

I'll look into it later, but any ideas welcome

**Comment by Y-Less** [comment]
I'll have to look more as well, but that looks like `OnScriptInit` is being redefined by ALS macros to `Prefix_OnScriptInit`.  y_hooks is actually meant to protect against this by spotting any callbacks that start with `On` and mangling the names in such a way that they are hidden from further replacements by the pre-processor.  I do the same for strings so that `#OnGameModeInit` is mutated to `"On"#GameModeInit` then it can't be detected and replaced with (say) `#AMX_OnGameModeInit`.  However, clearly something isn't working...

**Comment by Y-Less** [comment]
The first error was caused by I don't know what - you are apparently right in that short-circuiting didn't work:

```pawn
#if defined _DEBUG
	#if _DEBUG >= 1
		#define y_lock%0; _DoYLock(%0);
	#else
		#define y_lock%0; _DoYLock();
	#endif
#else
	#define y_lock%0; _DoYLock();
#endif
```

The second error was because of cyclical includes in YSI - normally they are quite hierarchical (or I try and keep them that way), but y_lock spreads code all over the place to make itself hard to track down and modify (security by obscurity, the worst kind I know), so it has some code in low-level libraries that actually rely on higher-level libraries.  In this case, the code was in y_shortfunc, an internal library that gives very short names to some common functions for the purposes of keeping macro contents small.  It is used by a number of libraries, and while I don't think the functions are used by y_hooks directly (since it doesn't have complex macros that need shrinking) it is still indirectly included by one of y_hooks dependencies.  As a result, y_hooks is included, which includes y_shortfunc as part of its preamble, before getting to defining the macros.  Then y_shortfunc includes y_hooks to get the macro definitions, which creates a cycle.  Cycles aren't bad on their own - thats partially what include guards are there to prevent against, but in this case the macro hadn't been defined yet and the guard stopped it being defined by the second include.

Long story short, do this:

```pawn
	forward UNIQUE_FUNCTION<@yH_OnScriptInit@...>();
	
	public UNIQUE_FUNCTION<@yH_OnScriptInit@...>()
```

Instead of the hook on line 519.

**Comment by Y-Less** [comment]
And add this to the bottom of YSI_Internal\y_unique.inc:

```pawn
#if !defined Y_UNIQUE_LOCK_CONTRIBUTION
	#define Y_UNIQUE_LOCK_CONTRIBUTION 0
#endif

#if defined YSI_LOCK_MODE
	#if Y_UNIQUE_LOCK_CONTRIBUTION < 2
		forward UNIQUE_FUNCTION<@yH_OnScriptInit@...>();
		
		public UNIQUE_FUNCTION<@yH_OnScriptInit@...>()
		{
			#if Y_UNIQUE_LOCK_CONTRIBUTION == 0
				#undef Y_UNIQUE_LOCK_CONTRIBUTION
				#define Y_UNIQUE_LOCK_CONTRIBUTION 1
				YSI_gLockData[1] -= 'i' - '.';
			#else
				#undef Y_UNIQUE_LOCK_CONTRIBUTION
				#define Y_UNIQUE_LOCK_CONTRIBUTION 2
				YSI_gLockData{11} = 0;
			#endif
		}
		
		#include "y_unique"
	#endif
#endif
```

**Comment by Y-Less** [comment]
Also, if you want another way to secure your server, there's a pretty cool new library that can take input from the server at startup:

http://forum.sa-mp.com/showthread.php?t=624230

Code could be improved, but this would combine the two:

```pawn
//#define OD_SERVER_PASSWORD // Need to generate this (run the server once without a password).
#include <YSI_Server\y_lock>
#include <od_server_password>

public OnGameModeInit()
{
	OD_STDIO_OnGameModeInit(); // Not hooked correctly.
	if (OD_PASSWORD_OnGameModeInit())
	{
		y_lock 10;
	}
}

public OnGameModeExit()
{
	OD_STDIO_OnGameModeExit(); // Not hooked correctly.
}
```

With the changes I put above, that will force your server to a single IP and require a password, but there are some small tweaks still required because y_lock hasn't been fully ported (it will work, but needs improvement).
